includes/integration.php

```diff
@@ -0,0 +1,537 @@
+            $this->add_tag_to_every_customer = $this->get_option('add_tag_to_every_customer', '');
+            $this->prefix_order_ids = $this->get_option('prefix_order_ids', '');
+            $this->http_or_https = $this->get_option('http_or_https', 'https') == 'https' ? 'https' : 'http';
+            $this->accept_tracking = true;
+            $this->tracking_library = $this->initialize_metrilo_library();
+            
+            // initiate woocommerce hooks and activities
+            add_action('woocommerce_init', array($this, 'on_woocommerce_init'));
+            
+            // hook to integration settings update
+            add_action('woocommerce_update_options_integration_' . $this->id, array($this, 'process_admin_options'));
+        }
+        
+        public function on_woocommerce_init()
+        {
+            $this->load_helpers_and_data_models();
+            
+            // check if I should clear the events cookie queue
+            $this->check_for_metrilo_clear();
+            
+            // check if API token and Secret are both entered
+            $this->check_for_keys();
+            
+            // hook to WooCommerce models
+            $this->ensure_hooks();
+            
+            // ensure identification
+            $this->ensure_identify();
+            
+            // process cookie events
+            $this->process_cookie_events();
+        }
+        
+        public function load_helpers_and_data_models()
+        {
+            $this->activity_helper = include_once(METRILO_ANALYTICS_PLUGIN_PATH . 'helpers/class.metrilo-activity.php');
+            $this->api_client = include_once(METRILO_ANALYTICS_PLUGIN_PATH . 'helpers/class.metrilo-api-client.php');
+            
+            $this->customer_data = include_once(METRILO_ANALYTICS_PLUGIN_PATH . 'models/class.metrilo-customer-data.php');
+            $this->customer_serializer = include_once(METRILO_ANALYTICS_PLUGIN_PATH . 'helpers/class.metrilo-customer-serializer.php');
+            
+            $this->category_data = include_once(METRILO_ANALYTICS_PLUGIN_PATH . 'models/class.metrilo-category-data.php');
+            $this->category_serializer = include_once(METRILO_ANALYTICS_PLUGIN_PATH . 'helpers/class.metrilo-category-serializer.php');
+            
+            $this->product_data = include_once(METRILO_ANALYTICS_PLUGIN_PATH . 'models/class.metrilo-product-data.php');
+            $this->product_serializer = include_once(METRILO_ANALYTICS_PLUGIN_PATH . 'helpers/class.metrilo-product-serializer.php');
+            
+            $this->order_data = include_once(METRILO_ANALYTICS_PLUGIN_PATH . 'models/class.metrilo-order-data.php');
+            $this->order_serializer = include_once(METRILO_ANALYTICS_PLUGIN_PATH . 'helpers/class.metrilo-order-serializer.php');
+        }
+        
+        public function check_for_keys()
+        {
+            if (is_admin()) {
```

Comment:
I see that you are putting the entire code block of a function inside if here and there. You can use the so called guard clauses to reduce the amount of if/elses in your code.

In this function, you can go with
```php
if (!is_admin()) {
  return;
  }
  
 // continue with your code here
```

Correct me if I'm wrong, but this entire method could be turned to this:
```php
 public function check_for_keys() {
    if (!is_admin()) {
        return;
    }

    if ((empty($this->api_token) || empty($this->api_secret)) && empty($_POST['save'])) {
        add_action('admin_notices', array($this, 'admin_keys_notice'));
    }
    
    if (empty($_POST['save'])) {
        return;
    }

    $key = trim($_POST['woocommerce_metrilo-analytics_api_token']);
    $secret = trim($_POST['woocommerce_metrilo-analytics_api_secret']);
    
    if (empty($key) || empty($secret)) {
        return;
    }

    $response = $this->activity_helper->create_activity('integrated');
        
    if ($response) {
        add_action('admin_notices', array($this, 'admin_import_invite'));
    } else {
        WC_Admin_Settings::add_error($this->admin_import_error_message());
    }
}
```
and probably it could edited even further.

Reply:
This is something new, but I will watch for it when refactoring. Thumbs up for the suggestion.